simd/src/Kokkos_SIMD.hpp

```diff
@@ -99,83 +110,143 @@ struct ForSpace;
+  using type = host_fixed_native;
+
+  template <int N>
+  using simd_abi = host_native_abi<N>;
 };
 #endif
 
 #ifdef KOKKOS_ENABLE_CUDA
 template <>
 struct ForSpace<Kokkos::Cuda> {
   using type = scalar;
+
+  template <int N>
+  using simd_abi = scalar;
 };
 #endif
 
 #ifdef KOKKOS_ENABLE_THREADS
 template <>
 struct ForSpace<Kokkos::Threads> {
-  using type = host_native;
+  using type = host_fixed_native;
+
+  template <int N>
+  using simd_abi = host_native_abi<N>;
 };
 #endif
 
 #ifdef KOKKOS_ENABLE_HPX
 template <>
 struct ForSpace<Kokkos::Experimental::HPX> {
   using type = scalar;
+
+  template <int N>
+  using simd_abi = scalar;
 };
 #endif
 
 #ifdef KOKKOS_ENABLE_OPENMP
 template <>
 struct ForSpace<Kokkos::OpenMP> {
-  using type = host_native;
+  using type = host_fixed_native;
+
+  template <int N>
+  using simd_abi = host_native_abi<N>;
 };
 #endif
 
 #ifdef KOKKOS_ENABLE_OPENMPTARGET
 template <>
 struct ForSpace<Kokkos::Experimental::OpenMPTarget> {
   using type = scalar;
+
+  template <int N>
+  using simd_abi = scalar;
 };
 #endif
 
 #ifdef KOKKOS_ENABLE_OPENACC
 template <>
 struct ForSpace<Kokkos::Experimental::OpenACC> {
   using type = scalar;
+
+  template <int N>
+  using simd_abi = scalar;
 };
 #endif
 
 #ifdef KOKKOS_ENABLE_HIP
 template <>
 struct ForSpace<Kokkos::HIP> {
   using type = scalar;
+
+  template <int N>
+  using simd_abi = scalar;
 };
 #endif
 
 #ifdef KOKKOS_ENABLE_SYCL
 template <>
 struct ForSpace<Kokkos::SYCL> {
   using type = scalar;
+
+  template <int N>
+  using simd_abi = scalar;
 };
 #endif
 
+template <class Space = Kokkos::DefaultExecutionSpace>
+using native_fixed_abi = typename ForSpace<Space>::type;
+
+template <int N, class Space = Kokkos::DefaultExecutionSpace>
+using native_abi = typename ForSpace<Space>::template simd_abi<N>;
+
 }  // namespace Impl
 
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
 template <class Space>
 using ForSpace = typename Impl::ForSpace<typename Space::execution_space>::type;
 
 template <class T>
 using native = ForSpace<Kokkos::DefaultExecutionSpace>;
+#endif
 
 }  // namespace simd_abi
 
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
 template <class T>
-using native_simd = simd<T, simd_abi::native<T>>;
+using native_simd KOKKOS_DEPRECATED_WITH_COMMENT(
+    "Temporarily use Impl::simd<T> instead") =
+    basic_simd<T, simd_abi::native<T>>;
 template <class T>
-using native_simd_mask = simd_mask<T, simd_abi::native<T>>;
+using native_simd_mask KOKKOS_DEPRECATED_WITH_COMMENT(
+    "Temporarily use Impl::simd_mask<T> instead") =
+    basic_simd_mask<T, simd_abi::native<T>>;
+
+template <class T, class Abi>
+using simd KOKKOS_DEPRECATED_WITH_COMMENT(
+    "Temporarily use Impl::simd<T> instead") = basic_simd<T, Abi>;
+
+template <class T, class Abi>
+using simd_mask KOKKOS_DEPRECATED_WITH_COMMENT(
+    "Temporarily use Impl::simd_mask<T> instead") = basic_simd_mask<T, Abi>;
+#endif
 
 namespace Impl {
 
+template <class T, int N = 0>
+using simd =
```

Comment:
I don't quite get why these are in Impl?

Reply:
This is the part where I'm looking for some more opinions from reviewers.

In P1928, `simd<T, Abi>` is removed and `simd<T, N = native-size>` became an alias for `basic_simd<T, Abi>`.
However, in Kokkos simd, replacing `simd<T, Abi>` with `simd<T, N = native-size>` first needs to go through a deprecation. Until `simd<T, Abi>` is fully deprecate and removed, I thought `simd<T, N = native-size>` could be introduced in the `Impl` namespace temporarily and later be moved out of impl.

Reply:
I think that's ok.

Reply:
I would prefer we do not promote people actively using anything that is in `Impl`. Is there no other way?

Reply:
I guess one question is how many people would we be actively breaking here if we didn't put this in Impl? Aren't our customers still using an old version of this (and anyway it's in Experimental)? I guess it's some of the same arguments as removing reference

Reply:
> I would prefer we do not promote people actively using anything that is in `Impl`. Is there no other way?

I think an alternative approach could be to name the new alias to be something else temporarily and place it in the `Kokkos::Experimental` namespace. Then, once `simd<T, Abi>` is fully deprecated and removed, introduce `simd<T, N>`, followed by the deprecation of the temporary alias.

> I guess one question is how many people would we be actively breaking here if we didn't put this in Impl? Aren't our customers still using an old version of this (and anyway it's in Experimental)?

I think the most common use of Kokkos simd is through `native_simd<...>` (which is being deprecated here), but it's still been valid to use `simd<T, Abi>` for specifying simd vector width.
Although very tempting, I'm a bit hesitant to skip the deprecation step and replace `simd<T, Abi>` with `simd<T, N>` right away.

Reply:
> it's still been valid to use `simd<T, Abi>` for specifying simd vector width. Although very tempting, I'm a bit hesitant to skip the deprecation step and replace `simd<T, Abi>` with `simd<T, N>` right away.

hmm ... as it is still in `Experimental` I would actually prefer to make it compliant with the P1928 and not have a state where we tell people to use something out of `Impl`. Especially since it would break user code twice.

Reply:
Removed `simd<T, Abi>` from `Kokkos::Experimental` namespace and replaced it with `simd<T, N>` without deprecation as discussed.

Reply:
Hmm ... can we add some `simd` and `simd_mask` in some tests